File: Microsoft-prototype/filterSpeech.cpp
#include <cctype>
#include "filterSpeech.h"
#include "Data.h"
#include "Feature.h"
#include <iostream>
#include "algorithm"
using namespace std;

FilterSpeech::FilterSpeech(const SurgeryType& procedure)
	: procedure_(procedure)
{}

const std::vector<shared_ptr<Tool>>& FilterSpeech::setToolList() const {
	if (procedure_ == SurgeryType::Trauma) {
		return toolListTrauma;
	}
	if (procedure_ == SurgeryType::Spine) {
		return toolListSpine;
	}
	if (procedure_ == SurgeryType::Reconstruction) {
		return toolListReconstruction;
	}
	else {
		return std::vector<shared_ptr<Tool>>();
	}
}

void FilterSpeech::searchProperties(const string& speechTranscription, ToolPropertySearch& toolPropertySearch)const {
	const string& speechTranscriptionLower = convertToLowerCase(speechTranscription);
	const std::vector<shared_ptr<Tool>>& toolList = setToolList();
	for (const auto& tool : toolList) {
		if (!toolPropertySearch.isToolFound) {
			tool->searchProperty(speechTranscriptionLower, toolPropertySearch);
		}
	}
}


void FilterSpeech::reinitPropertySearch(ToolPropertySearch& toolPropertySearch)const
{
	toolPropertySearch.isToolFound = false;
	toolPropertySearch.toolHandle = ToolHandleType::BadToolHandle;
	if (toolPropertySearch.screwSpinePropertySearch.diameter != DiameterScrew::BadDiameter && toolPropertySearch.screwSpinePropertySearch.length != Length::BadLength) {

		toolPropertySearch.screwSpinePropertySearch.diameter = DiameterScrew::BadDiameter;
		toolPropertySearch.screwSpinePropertySearch.length = Length::BadLength;
	}
	if (toolPropertySearch.screwWristPropertySearch.length != Length::BadLength &&
		toolPropertySearch.screwWristPropertySearch.headType!= HeadType::BadHeadType &&
		toolPropertySearch.screwWristPropertySearch.threadFullness!= ThreadFullness::BadThreadFulness) {
		toolPropertySearch.screwWristPropertySearch.length = Length::BadLength;
		toolPropertySearch.screwWristPropertySearch.threadFullness = ThreadFullness::BadThreadFulness ;
		toolPropertySearch.screwWristPropertySearch.headType = HeadType::BadHeadType;
	}
	if (toolPropertySearch.length!= Length::BadLength) {
		toolPropertySearch.length = Length::BadLength;
	}

	if (toolPropertySearch.screwAnklePropertySearch.threadType != ThreadType::BadThreadType && toolPropertySearch.screwAnklePropertySearch.length != Length::BadLength) {
		toolPropertySearch.screwAnklePropertySearch.threadType = ThreadType::BadThreadType;
		toolPropertySearch.screwAnklePropertySearch.length = Length::BadLength;
	}
}


string FilterSpeech::otherFeature(const string& result)const {
	for (const Feature& feature : otherFeatures) {
		for (const auto& name : feature.featureNames_) {
			if (result.find(string(name)) != string::npos) {
				return feature.action_;
			}
		}

	}
	return "";
}

string FilterSpeech::convertToLowerCase(const string& InString) const {
	std::string result(InString);
	std::transform(std::cbegin(result), std::cend(result), std::begin(result),
		[](unsigned char c) -> unsigned char { return std::tolower(c); });
	return result;
}

File: vosk-prototype/filterSpeech.py
from SurgeryType import SurgeryType
import Data
from ToolHandleType import ToolHandleType
from ToolProperties import Length
from ToolProperties import HeadType
from ToolProperties import DiameterScrew
from ToolProperties import ThreadFullness
from ToolProperties import ThreadType


def display_speech(text):
    print(text)


class FilterSpeech:
    def __init__(self, surgeryType):
        self.surgeryType_ = surgeryType

    def settoollist(self):
        if self.surgeryType_ == SurgeryType.Spine.value:
            return Data.toolListSpine
        elif self.surgeryType_ == SurgeryType.Trauma.value:
            return Data.toolListTrauma
        elif self.surgeryType_ == SurgeryType.Reconstruction.value:
            return Data.toolListReconstruction

    def search_properties(self, speech_transcription, tool_property_search):
        tool_list = self.settoollist()
        for tool in tool_list:
            if not tool_property_search.is_tool_found_:
                tool.search_property(speech_transcription, tool_property_search)
        self.search_features(speech_transcription, tool_property_search)

    def search_features(self, speech_transcription, tool_property_search):
        for feature in Data.otherFeatures:
            for name in feature.feature_names_:
                if speech_transcription.find(name) != -1:
                    tool_property_search.feature_action_ = feature.action_
                    tool_property_search.is_feature_found_ = True

    def reinit_property_search(self, tool_property_search):
        tool_property_search.is_tool_found_ = False
        tool_property_search.tool_handle_ = ToolHandleType.BadToolHandle
        if tool_property_search.length_ != Length.BadLength:
            tool_property_search.length_ = Length.BadLength
        if tool_property_search.screw_spine_property_search_.diameter_ != DiameterScrew.BadDiameter and \
                tool_property_search.screw_spine_property_search_.length_ != Length.BadLength:
            tool_property_search.screw_spine_property_search_.diameter_ = DiameterScrew.BadDiameter
            tool_property_search.screw_spine_property_search_.length_ = Length.BadLength
        if tool_property_search.screw_wrist_property_search_.length_ != Length.BadLength and \
                tool_property_search.screw_wrist_property_search_.head_type_ != HeadType.BadHeadType and \
                tool_property_search.screw_wrist_property_search_.thread_fullness_ != ThreadFullness.BadThreadFulness:
            tool_property_search.screw_wrist_property_search_.length_ = Length.BadLength
            tool_property_search.screw_wrist_property_search_.head_type_ = HeadType.BadHeadType
            tool_property_search.screw_wrist_property_search_.thread_fullness_ = ThreadFullness.BadThreadFulness
        if tool_property_search.screw_ankle_property_search_.thread_type_ != ThreadType.BadThreadType and \
                tool_property_search.screw_ankle_property_search_.length_ != Length.BadLength:
            tool_property_search.screw_ankle_property_search_.thread_type_ = ThreadType.BadThreadType
            tool_property_search.screw_ankle_property_search_.length_ = Length.BadLength
        if tool_property_search.feature_action_:
            tool_property_search.feature_action_ = ""
        if tool_property_search.feed_back_user_:
            tool_property_search.feed_back_user_ = ""
